File: core/application/service/work/Activity.go
// SPDX-License-Identifier: BUSL-1.1
// Copyright (c) 2026 Jérôme Bastian Winkel

package work

import (
	"context"
	"time"

	activityrepo "github.com/Jersyfi/hubtask/core/application/repository/activity"
	appshared "github.com/Jersyfi/hubtask/core/application/shared"
	"github.com/Jersyfi/hubtask/core/domain/model/activity"
	"github.com/Jersyfi/hubtask/core/domain/model/shared"
	domain "github.com/Jersyfi/hubtask/core/domain/model/work"
	"github.com/Jersyfi/hubtask/core/port/clock"
)

// The fourth thing a change to an entry owes, in one place.
//
// The other three are already shared - the event outwards, the change log for offline clients, the
// audit entry - and each writer records them itself. This one is a value every writer holds rather
// than a method each of them repeats, because what an entry's history says about a change must not
// depend on which use case made it: twelve copies of "build the entry, take an identifier, stamp
// the actor" is twelve chances for one of them to record the actor of the request instead of the
// actor of the change.

// ActivityJournal writes the steps of an item's own history (domain-model.md §3.5).
type ActivityJournal struct {
	Entries activityrepo.Journal
	IDs     clock.IDGenerator
}

// record writes one step, inside the caller's transaction.
//
// It fails the transaction when there is nowhere to write rather than carrying on quietly. A change
// whose history is missing one step is a history nobody can trust afterwards, and unlike an event or
// an audit entry there is no second record from which the gap could later be noticed.
func (j ActivityJournal) record(
	ctx context.Context, actor appshared.ActorContext, item domain.WorkItem,
	verb activity.Verb, changeSet map[string]any, at time.Time,
) error {
	if j.Entries == nil {
		return shared.ErrInternal.
			WithDetail("activity.journal_missing").
			WithParams(map[string]string{"verb": string(verb)})
	}

	return j.Entries.Record(ctx, activity.Entry{
		ID:       j.IDs.NewID(),
		TenantID: item.TenantID,
		ItemID:   item.ID,
		// The collection as it is *after* the change, which is the one the entry is now read
		// through. A move records where it went; where it came from is in the change set.
		CollectionID: item.CollectionID,
		Actor:        activity.Actor{Kind: actor.Kind, ID: actor.AccountID},
		Verb:         verb,
		ChangeSet:    changeSet,
		OccurredAt:   at,
	})
}

// verbIsTheChange is the change set of a step that moved no field.
//
// Archiving an entry is the whole of what happened, and a diff of the stamp it wrote would say the
// same thing a second time in a worse way. Written as a named function rather than an empty map
// literal at seven call sites, so that "this verb has no detail" is a decision a reader can see.
func verbIsTheChange() map[string]any { return map[string]any{} }

// historyForm is the form this item type's history takes (domain-model.md §2).
func historyForm(profile domain.CapabilityProfile) activity.Form {
	if profile.HistoryIsCompact() {
		return activity.Compact
	}
	return activity.Full
}

// historyFields turns the fields a change moved into the fields the history is asked to keep.
//
// The rule is the task's wording: the field names, and "content only where the product needs it".
// A title is content the product needs - "renamed Milk to Oat milk" is the entry somebody reads -
// and a note is content it does not: a note can be a page of text, and what its history is about is
// that somebody edited it. Everything else is an identifier, a rank or a state and is not content
// at all.
//
// A new content field is a decision to take here. The default is the values, which is right for the
// identifiers and states that make up most of them, and wrong for text - so a field carrying text
// belongs in the case above the default the day it arrives.
func historyFields(changes []domain.FieldChange) []activity.Field {
	fields := make([]activity.Field, 0, len(changes))

	for _, change := range changes {
		fields = append(fields, activity.Field{
			Name:   change.Field,
			Detail: historyDetailOf(change.Field),
			From:   change.From,
			To:     change.To,
		})
	}
	return fields
}

func historyDetailOf(field string) activity.Detail {
	switch field {
	case domain.FieldNotes:
		return activity.NameOnly
	default:
		return activity.WithValues
	}
}

File: apps/webapp/src/lib/data/activity.ts
// SPDX-License-Identifier: BUSL-1.1
// Copyright (c) 2026 Jérôme Bastian Winkel

/**
 * An entry's history, turned from a record into the parts of a sentence.
 *
 * The rule that shapes all of it is `domain-model.md` §3.5: **`verb` is a code**. `item.completed`
 * is stored, `activity.item_completed` is what a client renders, and a module that wrote
 * "Completed" would be the message catalogue growing a second copy — which ADR-0011 forbids and
 * F1-07 built the renderer to prevent. So nothing here produces text; it produces codes and
 * parameters, and `messages.t` turns them into words.
 *
 * Two model facts this must not smooth over.
 *
 * **The change set keeps field names always and values only where the product needs them.** A
 * rename carries both titles; a note carries `changed: true` and none of its text, because no user
 * content goes anywhere it is not needed (ADR-0017). `changesOf` reads `from`, `to` and `changed`
 * and **nothing else** — so a field the server one day carries more about does not leak through a
 * client that was written before it.
 *
 * **An activity's history is compact.** The verb, the actor and the time are the whole of the
 * step and the change set is empty, so there is nothing here that invents a detail for one.
 */

import type { ActivityEntry } from '@hubtask/sync-engine';

/** What the history kept about one field. */
export interface Change {
  readonly field: string;
  /** Where there was a value before. Absent when there was none — the contract's own wording. */
  readonly from?: string;
  readonly to?: string;
  /** True for a field whose values the history does not keep. A note is the worked example. */
  readonly isOpaque: boolean;
}

/** How a value is written when it is shown at all. Never an object, never JSON at a reader. */
function textOf(value: unknown): string | undefined {
  if (value === null || value === undefined) return undefined;
  if (typeof value === 'string') return value;
  if (typeof value === 'number' || typeof value === 'boolean') return String(value);
  // Anything else is a shape this client was not written for. It is left out rather than printed:
  // `[object Object]` in a history is worse than a field with no detail beside it.
  return undefined;
}

/**
 * What moved, as the history recorded it.
 *
 * The order is the change set's own. A client that sorted would be claiming an order the record
 * does not have, and the record's is at least the server's.
 */
export function changesOf(changeSet: Record<string, unknown> | undefined): readonly Change[] {
  if (!changeSet) return [];

  const changes: Change[] = [];
  for (const [field, moved] of Object.entries(changeSet)) {
    if (moved === null || typeof moved !== 'object') continue;
    const detail = moved as { from?: unknown; to?: unknown; changed?: unknown };
    changes.push({
      field,
      from: textOf(detail.from),
      to: textOf(detail.to),
      isOpaque: detail.changed === true,
    });
  }
  return changes;
}

/**
 * The change-set fields whose values are people rather than text.
 *
 * The four verbs C-01 added — `item.assigned`, `item.unassigned`, `item.member_added`,
 * `item.member_removed` — carry an account identifier where every other change carries a value a
 * reader can read. Showing the identifier would be showing a UUID to somebody asking who took the
 * entry over, so the name is looked up and the field list says which ones to look up.
 *
 * A list rather than a guess at "anything ending in `_id`": a `bucket_id` is not a person, and a
 * client that resolved every identifier as an account would ask for accounts that do not exist.
 */
const PEOPLE_FIELDS = new Set(['assignee_id', 'member_id', 'owner_id', 'actor_id']);

export function namesPeople(field: string): boolean {
  return PEOPLE_FIELDS.has(field);
}

/**
 * The change-set fields whose values are files.
 *
 * `item.attachment_added` and `item.attachment_removed` carry `media_id`, and a UUID in front of
 * somebody asking which file was attached is the same non-answer an account identifier is. The
 * name is one request away — the media record the client already reads for a cover — so the field
 * list says which values to look up, exactly as it does for people.
 *
 * `cover` is deliberately not here: it carries the whole cover object rather than an identifier,
 * and `textOf` already refuses to print an object at a reader.
 */
const MEDIA_FIELDS = new Set(['media_id']);

export function namesMedia(field: string): boolean {
  return MEDIA_FIELDS.has(field);
}

/** Every media identifier a set of changes mentions, each once. */
export function mediaNamedBy(changes: readonly Change[]): readonly string[] {
  const ids = new Set<string>();
  for (const change of changes) {
    if (!namesMedia(change.field)) continue;
    if (change.from) ids.add(change.from);
    if (change.to) ids.add(change.to);
  }
  return [...ids];
}

/** Every account identifier a set of changes mentions, each once. What the name lookup is asked. */
export function accountsNamedBy(changes: readonly Change[]): readonly string[] {
  const ids = new Set<string>();
  for (const change of changes) {
    if (!namesPeople(change.field)) continue;
    if (change.from) ids.add(change.from);
    if (change.to) ids.add(change.to);
  }
  return [...ids];
}

/**
 * The message codes that name who did it, most specific first.
 *
 * A list rather than one code, for the reason `problem.ts` keeps one: the actor kinds are an enum
 * in the contract and `domain-model.md` §2's rule is that such sets grow with the installation, so
 * a kind this client has never heard of falls back to the one sentence that is true of every
 * actor — "somebody" — rather than rendering a key.
 *
 * **These are the fallbacks, not the whole answer.** The contract says of the actor that "the
 * label is not here: the account is one request away", and `GET /accounts/{accountId}` is that
 * request — `lib/data/accounts.svelte.ts` makes it and caches what comes back. What this function
 * answers is what to say when no name was resolved: for the reader, "You"; for anybody else, the
 * sentence true of their kind. It invents no names, and it never has.
 */
export function actorCodes(entry: ActivityEntry, selfAccountId: string | undefined): string[] {
  const kind = entry.actor?.type;
  const id = entry.actor?.id;

  if (kind === 'USER' && id && id === selfAccountId) return ['app.activity.actor_you'];
  return [`app.activity.actor_${kind ?? 'UNKNOWN'}`, 'app.activity.actor_someone'];
}
